File: 2015/06/06.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

inst = []

File.readlines("./input/06.txt").each do |line|
  op = line[/(^toggle|^turn \w+)/]

  start, stop =
    line.scan(/\d+,\d+/)
        .map { |x| x.split(",").map(&:to_i) }

  inst << [op, start, stop]
end

def part1(inst)
  lights = Array.new(1000) { Array.new(1000) { 0 } }

  inst.each do |op, start, stop|
    lop = case op
      when "turn on"  then ->(x) { 1 }
      when "turn off" then ->(x) { 0 }
      when "toggle"   then ->(x) { 1 ^ x }
      end

    (start.first..stop.first).each do |y|
      (start.last..stop.last).each do |x|
        lights[y][x] = lop.(lights[y][x])
      end
    end
  end

  lights.sum &:sum
end

def part2(inst)
  lights = Array.new(1000) { Array.new(1000) { 0 } }

  inst.each do |op, start, stop|
    lop = case op
      when "turn on"  then ->(x) { x += 1 }
      when "turn off" then ->(x) { x -= 1 }
      when "toggle"   then ->(x) { x += 2 }
      end

    (start.first..stop.first).each do |y|
      (start.last..stop.last).each do |x|
        lights[y][x] = lop.(lights[y][x])
      end
    end
  end

  lights.sum &:sum
end

silver = part1(inst)
gold   = part2(inst)

puts(
  "Day 06\n"       \
  "======\n"       \
  "✮: #{silver}\n" \
  "★: #{gold}"
)

File: 2015/06/06.clj
(require '[clojure.string :refer (split-lines split)])

(set! *warn-on-reflection* true)
(set! *unchecked-math* true)

(defn to-inst [op start stop]
  [op
   (mapv #(Long. ^String %) (split start #","))
   (mapv #(Long. ^String %) (split stop #","))])

(defn parse [line]
  (->> line
       (re-find #"^(toggle|turn \w+) (\d+,\d+) .* (\d+,\d+)")
       rest
       (apply to-inst)))

(def ops {"turn on" (fn [^Long _] 1)
          "turn off" (fn [^Long _] 0)
          "toggle" (fn [^Long x] (bit-xor 1 x))})

(def ops2 {"turn on" (fn [^Long x] (inc x))
           "turn off" (fn [^Long x] (dec x))
           "toggle" (fn [^Long x] (+ 2 x))})

(defn transform! [^ints grid [op-str start end] part2]
  (let [opfn (if part2 (ops2 op-str) (ops op-str))]
    (doseq [^Long i (range (first start) (inc (first end)))
            ^Long j (range (second start) (inc (second end)))]
      (aset-int grid (+ (* 1000 i) j)
                (opfn (aget ^ints grid (+ (* 1000 i) j)))))))

(defn solve [instructions part2]
  (let [grid (int-array (* 1000 1000))]
    (doseq [inst instructions]
      (transform! grid inst part2))
    (areduce grid i cnt 0 (+ cnt (aget grid i)))))

(def instructions
  (->> "06.txt"
       slurp
       split-lines
       (mapv parse)))

(println "Day 06\n======")
(println (solve instructions false))
(println (solve instructions true))
